apis/python/src/tiledbsoma/options/soma_tiledb_context.py

```diff
@@ -29,3 +30,64 @@ class SOMATileDBContext:
     """
 
     tiledb_ctx: tiledb.Ctx = _build_default_tiledb_ctx()
+
+    read_timestamp: int = -1
+    """
+    Timestamp for operations on SOMA objects open in read mode, in milliseconds since Unix epoch.
+    Defaults to the time of context initialization. Set to 0xFFFFFFFFFFFFFFFF (UINT64_MAX) to get
+    the latest revision as of when *each* object is opened.
+
+    SOMA objects opened in *write* mode ignore any read timestamp.
+    """
+
+    read_timestamp_start: int = 0
+    """
+    Timestamp range start for operations on SOMA objects opened in read mode. This is usually zero
+    except for specific, unusual query requirements.
+    """
+
+    write_timestamp: Optional[int] = None
+    """
+    Timestamp applied to all SOMA object write operations. If unset, each individual write
+    operation receives the timestamp as of when the operation executes.
+
+    Caution: overlapping writes (of overlapping array slices, or setting the same collection key)
+    should be avoided when write_timestamp is set. Distinct, overlapping write operations given the
+    same timestamp may be applied in any order.
+    """
+
+    # (internal) tiledb.Ctx specifically for tiledb.Group operations; unlike arrays, tiledb.Group
+    # needs its timestamps set in the tiledb.Ctx.
+    _group_read_tiledb_ctx: tiledb.Ctx = attrs.field(init=False)
+    _group_write_tiledb_ctx: tiledb.Ctx = attrs.field(init=False)
+
+    def __attrs_post_init__(self) -> None:
+        "initialization hook invoked by the attrs-generated __init__"
+
+        if self.read_timestamp < 0:
+            object.__setattr__(self, "read_timestamp", int(time.time() * 1000))
```

Comment:
This might be better accomplished with a `factory` function:

```python
read_timestamp = attrs.field(type=int, factory=lambda: int(time.time() * 1000))
```